# **Pandas : Series, DataFrame 데이터의 처리**

# Pandas Library 설치 & Library Import

In [ ]:
!pip install pandas==1.5.3

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 66.2 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.10.1 requires pandas<2.2.3dev0,>=2.0, but you have pandas 1.5.3 which is incompatible.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 1.5.3 which is incompatible.
mizani 0.13.0 requires pandas>=2.2.0, but you have pandas 1.5.3 which is incompatible.
plotnine 0.14.1 requires pandas>=2.2.0, but you have pandas 1.5.3 which is incompatible.
xarray 2024.10.0 requires pandas>=2.1, but you have pandas 1.5.3 which is incompatible.


In [ ]:
import pandas as pd

# (1) Series

## Series 생성

List에서 Series 생성

In [ ]:
values = [20, 30, 40, 50]
series_1 = pd.Series(values)

In [ ]:
print(series_1)

0    20
1    30
2    40
3    50
dtype: int64


In [ ]:
series_1

0    20
1    30
2    40
3    50
dtype: int64

In [ ]:
list_stock = [92300, 94300, 92100, 92400, 92600]
stock = pd.Series(list_stock)

In [ ]:
print(stock)

0    92300
1    94300
2    92100
3    92400
4    92600
dtype: int64


Dictionary로 Series 생성하기


In [ ]:
dict_stock = {'2019-02-15': 92300,
              '2019-02-16': 94300,
              '2019-02-17': 92100,
              '2019-02-18': 92400,
              '2019-02-19': 92600}

In [ ]:
stock_series = pd.Series(dict_stock)

In [ ]:
stock_series

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
dtype: int64

## Series의 기초정보 확인하고 조작하기

In [ ]:
stock_series.index = ['2024-02-15', '2024-02-16', '2024-02-17', '2024-02-18', '2024-02-19']

In [ ]:
stock_series.index

Index(['2024-02-15', '2024-02-16', '2024-02-17', '2024-02-18', '2024-02-19'], dtype='object')

In [ ]:
type(stock_series.index)

pandas.core.indexes.base.Index

### index 조회
```python
series변수명.index
```

In [ ]:
# 딕셔너리 만들기
dict_stock = {'2019-02-15': 92300,
              '2019-02-16': 94300,
              '2019-02-17': 92100,
              '2019-02-18': 92400,
              '2019-02-19': 92600}
# 시리즈 만들기
stock = pd.Series(dict_stock)

In [ ]:
print(stock)

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
dtype: int64


In [ ]:
# 인덱스 확인
stock.index

Index(['2019-02-15', '2019-02-16', '2019-02-17', '2019-02-18', '2019-02-19'], dtype='object')

### index 변경
- series의 index 속성을 새로운 list로 변경/치환할 수 있습니다.
- 이때, series의 row개수와 새로운 list의 길이가 동일해야 에러가 나지않고 index변경이 가능합니다.

```python
series변수명.index = 새로운 index list
```

In [ ]:
new_index = ['2020-02-15', '2020-02-16', '2020-02-17', '2020-02-18', '2020-02-19']

In [ ]:
stock.index = new_index

In [ ]:
print(stock.index)

Index(['2020-02-15', '2020-02-16', '2020-02-17', '2020-02-18', '2020-02-19'], dtype='object')


### index 초기화(reset)

In [ ]:
stock_series.reset_index(drop =True, inplace = True)

In [ ]:
stock_series

0    92300
1    94300
2    92100
3    92400
4    92600
dtype: int64

### Series 데이터 타입 확인


In [ ]:
dict_stock = {'2019-02-15': 92300,
              '2019-02-16': 94300,
              '2019-02-17': 92100,
              '2019-02-18': 92400,
              '2019-02-19': 92600}
stock = pd.Series(dict_stock)

In [ ]:
stock.dtype

dtype('int64')

In [ ]:
dict_stock = {'2019-02-15': 92300.0,
              '2019-02-16': 94300.5,
              '2019-02-17': 92100.0,
              '2019-02-18': 92400.0,
              '2019-02-19': 92600.5}
stock = pd.Series(dict_stock)

In [ ]:
stock.dtype

dtype('float64')

### series 데이터타입 변경

In [ ]:
dict_stock = {'2019-02-15': 92300.0,
              '2019-02-16': 94300.5,
              '2019-02-17': 92100.0,
              '2019-02-18': 92400.0,
              '2019-02-19': 92600.5}
stock = pd.Series(dict_stock)

In [ ]:
stock.dtype

dtype('float64')

In [ ]:
stock = stock.astype("int")

In [ ]:
stock.dtype

dtype('int64')

### Series 모양 확인

In [ ]:
dict_stock = {'2019-02-15': 92300.0,
              '2019-02-16': 94300.5,
              '2019-02-17': 92100.0,
              '2019-02-18': 92400.0,
              '2019-02-19': 92600.5}
stock = pd.Series(dict_stock)

In [ ]:
stock.shape

(5,)

### Series 상위 5개행, 하위 5개행 미리보기

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock.shape

(10,)

In [ ]:
stock.head()

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
dtype: int64

In [ ]:
stock.tail()

2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
## 상위 7개행 조회
stock.head(7)

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
dtype: int64

### Series 기초 통계 확인

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
stock.describe()

count       10.000000
mean     92740.000000
std        839.576613
min      92100.000000
25%      92300.000000
50%      92400.000000
75%      92600.000000
max      94300.000000
dtype: float64

## Data Handling하기

Numpy Indexing Slicing 복습

In [ ]:
import numpy as np

my_arr = np.random.randint(1,100, (5,6))

In [ ]:
my_arr

array([[84, 85, 40, 38, 80, 10],
       [69, 79, 54, 75, 79, 77],
       [39, 67, 20, 24,  3, 24],
       [49, 33, 81, 53, 93, 25],
       [65, 80, 56, 66, 17, 13]])

In [ ]:
## indexing
my_arr[0]

array([84, 85, 40, 38, 80, 10])

In [ ]:
## slicing
my_arr[1:4]

array([[69, 79, 54, 75, 79, 77],
       [39, 67, 20, 24,  3, 24],
       [49, 33, 81, 53, 93, 25]])

In [ ]:
## slicing
my_arr[0, 1:4]

array([85, 40, 38])

### Indexing, slicing

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
stock['2019-02-21']

94300

In [ ]:
stock['2019-02-21': '2019-02-23']

2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
dtype: int64

#### loc indexing, slicing
- loc는 index명으로 indexing/slicing을 하는 타입

- loc indexing
```python
series변수명.loc[index명]
```

- loc slicing
```python
series변수명.loc[시작index명:끝index명]
```

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
stock.loc['2019-02-22': '2019-02-24']

2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
stock.loc['2019-02-17']

92100

In [ ]:
stock.loc['2019-02-17':'2019-02-20']

2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
dtype: int64

#### iloc indexing, slicing

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

In [ ]:
stock.iloc[1: 3]

2019-02-16    94300
2019-02-17    92100
dtype: int64

In [ ]:
stock.iloc[3]

92400

In [ ]:
stock.iloc[3:10]

2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
dtype: int64

조건에 따라 추출하기

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

In [ ]:
stock.loc[(stock> 92500) | (stock<92300)]

2019-02-16    94300
2019-02-17    92100
2019-02-19    92600
2019-02-21    94300
2019-02-22    92100
2019-02-24    92600
dtype: int64

In [ ]:
stock.loc[stock > 92500]

2019-02-16    94300
2019-02-19    92600
2019-02-21    94300
2019-02-24    92600
dtype: int64

In [ ]:
stock.loc[(stock > 92300) & (stock < 92600)]

2019-02-18    92400
2019-02-23    92400
dtype: int64

In [ ]:
stock.loc[(stock < 92300) | (stock > 92500)]

2019-02-16    94300
2019-02-17    92100
2019-02-19    92600
2019-02-21    94300
2019-02-22    92100
2019-02-24    92600
dtype: int64

### Data 추가/수정/삭제

In [ ]:
dates = ['2019-02-15', '2019-02-16','2019-02-17','2019-02-18','2019-02-19', '2019-02-20', '2019-02-21','2019-02-22','2019-02-23','2019-02-24']
stock = pd.Series([92300, 94300, 92100, 92400, 92600, 92300, 94300, 92100, 92400, 92600], index=dates)

Data 추가

In [ ]:
stock['2019-02-25'] = 90000

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92600
2019-02-25    90000
dtype: int64

 Data 수정

In [ ]:
stock['2019-02-24'] = 92000

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92000
2019-02-25    90000
dtype: int64

#### Data 삭제

In [ ]:
del stock['2019-02-25']

In [ ]:
stock

2019-02-15    92300
2019-02-16    94300
2019-02-17    92100
2019-02-18    92400
2019-02-19    92600
2019-02-20    92300
2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
2019-02-24    92000
dtype: int64

# (2) DataFrame

## Data Frame 생성하기(From Dictionary)

In [ ]:
import pandas as pd

In [ ]:
my_dict = {'이름': ['사나', '모모', '쯔위', '나연'],
        '등급': ['Gold', 'Bronze', 'Silver', 'Gold'],
        '점수': [56000, 23000, 44000, 52000]}

In [ ]:
df = pd.DataFrame(my_dict)

In [ ]:
df

이름      등급     점수
0  사나    Gold  56000
1  모모  Bronze  23000
2  쯔위  Silver  44000
3  나연    Gold  52000


## DataFrame 생성하기(From array)

In [ ]:
import numpy as np
arr = np.array([['사나', 'Gold', 56000],
               ['모모', 'Bronze', 23000],
               ['쯔위', 'Silver', 44000],
               ['나연', 'Gold', 52000]]
               )

In [ ]:
cols = ['name', 'level', 'score']

In [ ]:
import pandas as pd

df = pd.DataFrame(columns = cols, data = arr)

In [ ]:
df

name   level  score
0   사나    Gold  56000
1   모모  Bronze  23000
2   쯔위  Silver  44000
3   나연    Gold  52000

## DataFrame 생성하기(From csv file)

In [ ]:
cd /content/drive/MyDrive/SeSAC_도봉_AI_2기/chapter4.Data_Analysis/data

/content/drive/MyDrive/SeSAC_도봉_AI_2기/chapter4.Data_Analysis/data


In [ ]:
pwd

'/content/drive/MyDrive/SeSAC_도봉_AI_2기/chapter4.Data_Analysis/data'

In [ ]:
ls

churn_data.csv                seoul_public_bike_usage.csv
ds_salaries.csv               tpss_sta_route_hpasngr_2023.09.04-09.10.csv
kospi_data.csv                tpss_sta_route_hpasngr_2023.09.11-09.17.csv
real_estate_price.csv         tpss_sta_route_hpasngr_2023.09.18-09.24.csv
real_estate_price_edited.csv  tpss_sta_route_hpasngr_2023.09.25-10.01.csv
sales_info.csv                user_info.csv
seoul_public_bike_site.csv    자치구정류장코드분류.xlsx


In [ ]:
df = pd.read_csv('./real_estate_price.csv')

In [ ]:
df.tail()

시군구   번지   본번  부번              단지명  전용면적(㎡)    계약년월  계약일  \
7168  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  84.9907  202003   14   
7169  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438  202003   21   
7170  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  84.9907  202003   27   
7171  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438  202004   18   
7172  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438  202005    5   

            거래금액(만원)   층  건축년도      도로명  
7168         101,000   1  2009  서판교로 29  
7169         100,500   6  2009  서판교로 29  
7170         100,200  16  2009  서판교로 29  
7171          97,000   9  2009  서판교로 29  
7172          98,000   3  2009  서판교로 29



## DataFrame 기초 정보확인하고 조작하기

### DataFrame 미리보기(상위/하위 5개행)

In [ ]:
df.head()

Unnamed: 0            시군구  번지  본번  부번                단지명  전용면적(㎡)    계약년월  \
0           0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   
1           1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70.66  201906   
2           2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   
3           3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   
4           4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   

   계약일            거래금액   층  건축년도      도로명  
0    1          49,000   2  1995  미금로 184  
1    7          73,200  15  1995  미금로 184  
2   10          51,000   6  1995  미금로 184  
3   13          51,000   2  1995  미금로 184  
4   13          54,000   6  1995  미금로 184

In [ ]:
df.tail()

Unnamed: 0            시군구   번지   본번  부번              단지명  전용면적(㎡)  \
7168        7168  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  84.9907   
7169        7169  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438   
7170        7170  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  84.9907   
7171        7171  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438   
7172        7172  경기도 성남분당구 판교동  585  585   0  판교원마을9단지(한림풀에버)  71.8438   

        계약년월  계약일            거래금액   층  건축년도      도로명  
7168  202003   14         101,000   1  2009  서판교로 29  
7169  202003   21         100,500   6  2009  서판교로 29  
7170  202003   27         100,200  16  2009  서판교로 29  
7171  202004   18          97,000   9  2009  서판교로 29  
7172  202005    5          98,000   3  2009  서판교로 29

### index 조회

In [ ]:
df = pd.read_csv('./real_estate_price.csv')

In [ ]:
df.index

RangeIndex(start=0, stop=7173, step=1)

In [ ]:
list(df.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


### index 변경

In [ ]:
new_index = range(100, 7273)

In [ ]:
df.index = new_index

In [ ]:
df

시군구   번지   본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
100   경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906    1   
101   경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  70.6600  201906    7   
102   경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   10   
103   경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
104   경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
...             ...  ...  ...  ..                ...      ...     ...  ...   
7268  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   14   
7269  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202003   21   
7270  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   27   
7271  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202004   18   
7272  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202005    5   

            거래금액(만원)   층  건축년도      도로명  
100           49,000   2  1995  미금로 184  
101           73,200  15  1995  미금로 184  
102           51,000   6  1995  미금로 184  
103           51,000   2  1995  미금로 184  
104           54,000   6  1995  미금로 184  
...              ...  ..   ...      ...  
7268         101,000   1  2009  서판교로 29  
7269         100,500   6  2009  서판교로 29  
7270         100,200  16  2009  서판교로 29  
7271          97,000   9  2009  서판교로 29  
7272          98,000   3  2009  서판교로 29  

[7173 rows x 12 columns]

### index 초기화(reset)

In [ ]:
df.reset_index(inplace = True, drop = True)

In [ ]:
df

시군구   번지   본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
0     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906    1   
1     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  70.6600  201906    7   
2     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   10   
3     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
4     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
...             ...  ...  ...  ..                ...      ...     ...  ...   
7168  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   14   
7169  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202003   21   
7170  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   27   
7171  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202004   18   
7172  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202005    5   

            거래금액(만원)   층  건축년도      도로명  
0             49,000   2  1995  미금로 184  
1             73,200  15  1995  미금로 184  
2             51,000   6  1995  미금로 184  
3             51,000   2  1995  미금로 184  
4             54,000   6  1995  미금로 184  
...              ...  ..   ...      ...  
7168         101,000   1  2009  서판교로 29  
7169         100,500   6  2009  서판교로 29  
7170         100,200  16  2009  서판교로 29  
7171          97,000   9  2009  서판교로 29  
7172          98,000   3  2009  서판교로 29  

[7173 rows x 12 columns]

### DataFrame column 조회

In [ ]:
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적(㎡)', '계약년월', '계약일', '거래금액(만원)',
       '층', '건축년도', '도로명'],
      dtype='object')

### DataFrame Column명 변경

#### 방식1

In [ ]:
df = pd.read_csv('./real_estate_price.csv')

In [ ]:
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적(㎡)', '계약년월', '계약일', '거래금액(만원)',
       '층', '건축년도', '도로명'],
      dtype='object')

In [ ]:
new_cols = ['시군구', '번지', '본번', '부번', '단지명', '전용면적', '계약년월', '계약일', '거래금액','층', '건축년도', '도로명']

In [ ]:
df.columns = new_cols

In [ ]:
df.head()

시군구  번지  본번  부번                단지명   전용면적    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)  51.32  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)  70.66  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)  51.32  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)  51.32  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)  51.32  201906   13   

             거래금액   층  건축년도      도로명  
0          49,000   2  1995  미금로 184  
1          73,200  15  1995  미금로 184  
2          51,000   6  1995  미금로 184  
3          51,000   2  1995  미금로 184  
4          54,000   6  1995  미금로 184

#### 방식2

In [ ]:
df.rename(columns = {"거래금액(만원)": "거래금액"}, inplace = True)

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70.66  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   

             거래금액   층  건축년도      도로명  
0          49,000   2  1995  미금로 184  
1          73,200  15  1995  미금로 184  
2          51,000   6  1995  미금로 184  
3          51,000   2  1995  미금로 184  
4          54,000   6  1995  미금로 184

### DataFrame shape 확인

In [ ]:
df.shape

(7173, 12)

### DataFrmae Value확인

In [ ]:
df.values

array([['경기도 성남분당구 구미동', '77', 77, ..., 2, 1995, '미금로 184'],
       ['경기도 성남분당구 구미동', '77', 77, ..., 15, 1995, '미금로 184'],
       ['경기도 성남분당구 구미동', '77', 77, ..., 6, 1995, '미금로 184'],
       ...,
       ['경기도 성남분당구 판교동', '585', 585, ..., 16, 2009, '서판교로 29'],
       ['경기도 성남분당구 판교동', '585', 585, ..., 9, 2009, '서판교로 29'],
       ['경기도 성남분당구 판교동', '585', 585, ..., 3, 2009, '서판교로 29']],
      dtype=object)

In [ ]:
df.values[0]

array(['경기도 성남분당구 구미동', '77', 77, 0, '까치마을(1단지)(대우롯데선경)', 51.32, 201906,
       1, '        49,000', 2, 1995, '미금로 184'], dtype=object)

In [ ]:
df.values.shape

(7173, 12)

### DataFrame 데이터 타입 확인

In [ ]:
df.dtypes

시군구      object
번지       object
본번        int64
부번        int64
단지명      object
전용면적    float64
계약년월      int64
계약일       int64
거래금액     object
층         int64
건축년도      int64
도로명      object
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   시군구      7173 non-null   object 
 1   번지       7173 non-null   object 
 2   본번       7173 non-null   int64  
 3   부번       7173 non-null   int64  
 4   단지명      7173 non-null   object 
 5   전용면적(㎡)  7173 non-null   float64
 6   계약년월     7173 non-null   int64  
 7   계약일      7173 non-null   int64  
 8   거래금액     7173 non-null   object 
 9   층        7173 non-null   int64  
 10  건축년도     7173 non-null   int64  
 11  도로명      7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


### DataFrame 데이터 타입 변경

In [ ]:
df['계약년월'] = df['계약년월'].astype("str")

In [ ]:
df['전용면적'].astype("int")

0       51
1       70
2       51
3       51
4       51
        ..
7168    84
7169    71
7170    84
7171    71
7172    71
Name: 전용면적, Length: 7173, dtype: int64

In [ ]:
df['전용면적'] = df['전용면적'].astype("int")

In [ ]:
df.dtypes

시군구     object
번지      object
본번       int64
부번       int64
단지명     object
전용면적     int64
계약년월     int64
계약일      int64
거래금액    object
층        int64
건축년도     int64
도로명     object
dtype: object

### DataFrame column별 정보 확인

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   시군구     7173 non-null   object
 1   번지      7173 non-null   object
 2   본번      7173 non-null   int64 
 3   부번      7173 non-null   int64 
 4   단지명     7173 non-null   object
 5   전용면적    7173 non-null   int64 
 6   계약년월    7173 non-null   int64 
 7   계약일     7173 non-null   int64 
 8   거래금액    7173 non-null   object
 9   층       7173 non-null   int64 
 10  건축년도    7173 non-null   int64 
 11  도로명     7173 non-null   object
dtypes: int64(7), object(5)
memory usage: 672.6+ KB


In [ ]:
df_new = pd.read_csv('./real_estate_price_edited.csv')

In [ ]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    6939 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


### DataFrame 기초 통계 확인

In [ ]:
df.describe()

본번           부번      전용면적(㎡)          계약일            층  \
count  7173.000000  7173.000000  7173.000000  7173.000000  7173.000000   
mean    248.037781     0.030671    86.778428    15.578558     9.409034   
std     241.541311     0.241201    36.585154     8.785374     6.075077   
min       6.000000     0.000000    17.710000     1.000000     0.000000   
25%      91.000000     0.000000    58.800000     8.000000     5.000000   
50%     156.000000     0.000000    84.600000    15.000000     9.000000   
75%     308.000000     0.000000   101.980000    23.000000    13.000000   
max     989.000000     4.000000   244.580000    31.000000    36.000000   

              건축년도  
count  7173.000000  
mean   1996.712951  
std       6.177631  
min    1991.000000  
25%    1993.000000  
50%    1994.000000  
75%    1996.000000  
max    2015.000000

### DataFrame column의 Unique항목 조회

In [ ]:
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적(㎡)', '계약년월', '계약일', '거래금액', '층',
       '건축년도', '도로명'],
      dtype='object')

In [ ]:
df['단지명'].unique()

array(['까치마을(1단지)(대우롯데선경)', '까치마을(2단지)(주공)', '까치마을(3단지)(신원)',
       '까치마을(4단지)(롯데)', '까치마을(4단지)(선경)', '무지개(10단지)(건영)', '무지개(10단지)(삼성)',
       '무지개(11단지)(금강센테리움)', '무지개(12단지)(주공뜨란체)', '무지개(1단지)(대림)',
       '무지개(2단지)(엘지)', '무지개(3단지)(건영)', '무지개(3단지)(신한)', '무지개(4단지)(주공)',
       '무지개(5단지)(청구)', '무지개(6단지)(건영)', '무지개(7단지)(라이프)', '무지개(8단지)(제일)',
       '무지개(9단지)(동아)', '하얀마을(5단지)(주공)', '두산위브트레지움', '분당하우스토리', '삼라마이다스빌',
       '청솔마을(계룡)', '청솔마을(공무원)', '청솔마을(대원)', '청솔마을(동아)', '청솔마을(서광)',
       '청솔마을(성원)', '청솔마을(영남)', '청솔마을(유천)', '청솔마을(주공9단지)', '청솔마을(한라)',
       '코오롱더프라우', '백현마을2단지(주공)일부임대', '백현마을5단지(주공)', '백현마을6단지(주공)',
       '백현마을7단지(주공)', '백현마을8단지(대림)임대', '백현마을9단지(e편한세상)', '판교알파리움1단지',
       '판교알파리움2단지', '판교푸르지오그랑블', '샛별마을(동성)', '샛별마을(라이프)', '샛별마을(삼부)',
       '샛별마을(우방)', '장안타운(건영)', '봇들마을1단지(판교신미주)', '봇들마을2단지(이지더원)',
       '봇들마을3단지(주공)임대', '봇들마을4단지(주공)', '봇들마을7단지', '봇들마을8단지(주공)',
       '봇들마을9단지(금호어울림)', '판교푸르지오월드마크', '판교호반써밋플레이스', '시범삼성', '시범우성',
       '시범한신', '시범한양', '시범현대', '효자촌(대

### DataFrame 칼럼별 unique항목별로 count 집계하기

In [ ]:
df['단지명'].value_counts()

시범한양                 146
상록마을(우성)1            130
한솔마을(5단지)(주공)        128
한솔마을(4단지)(주공)        125
장미마을(현대)             124
                    ... 
판교원마을12단지(휴먼시아)임대      5
삼라마이다스빌                5
무지개(9단지)(동아)           5
성원상떼뷰101동              4
백현마을8단지(대림)임대          2
Name: 단지명, Length: 180, dtype: int64

### DataFrame 그룹별 통계 확인하기

In [ ]:
df['거래금액(만원)'].astype("int")

ValueError: invalid literal for int() with base 10: '        49,000'

In [ ]:
# df['거래금액(만원)'].str.replace(",", "").str.replace(" ", "")



df['거래금액(만원)'] = df['거래금액(만원)'].apply(lambda x: x.replace(",", "")).astype("int")

In [ ]:
df.select_dtypes("int").groupby(by = "계약년월", as_index = False).mean()

계약년월          본번        부번        계약일      거래금액(만원)          층  \
0   201906  212.530612  0.022263  15.801484  83376.808905  10.345083   
1   201907  222.456163  0.036849  16.162643  85526.238882   9.611182   
2   201908  255.876333  0.029851  17.916844  86186.251599   9.995736   
3   201909  240.095865  0.046992  16.592105  87299.124060   9.552632   
4   201910  225.369822  0.034517  15.714004  89325.041420   9.440828   
5   201911  236.480335  0.040167  16.037657  92478.555649   9.176569   
6   201912  267.266833  0.021197  13.208229  88400.487531   9.182045   
7   202001  275.924406  0.010799  14.708423  76500.805616   9.211663   
8   202002  259.686163  0.017118  15.744650  81288.786020   9.000000   
9   202003  357.147860  0.019455  13.311284  81381.038911   8.727626   
10  202004  317.593023  0.046512  17.005814  85040.720930   9.459302   
11  202005  248.504132  0.041322  14.020661  84969.012397   9.231405   

           건축년도  
0   1996.204082  
1   1996.097840  
2   1997.477612  
3   1996.796992  
4   1996.228797  
5   1996.805021  
6   1997.048628  
7   1996.390929  
8   1996.339515  
9   1998.731518  
10  1998.645349  
11  1996.822314

In [ ]:
df.groupby("계약년월").mean()

<ipython-input-128-9fccc4fb1ba3>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby("계약년월").mean()


본번        부번       전용면적        계약일          층         건축년도
계약년월                                                                      
201906  212.530612  0.022263  87.740260  15.801484  10.345083  1996.204082
201907  222.456163  0.036849  89.616264  16.162643   9.611182  1996.097840
201908  255.876333  0.029851  89.021322  17.916844   9.995736  1997.477612
201909  240.095865  0.046992  89.780075  16.592105   9.552632  1996.796992
201910  225.369822  0.034517  90.196252  15.714004   9.440828  1996.228797
201911  236.480335  0.040167  90.489540  16.037657   9.176569  1996.805021
201912  267.266833  0.021197  84.948878  13.208229   9.182045  1997.048628
202001  275.924406  0.010799  75.820734  14.708423   9.211663  1996.390929
202002  259.686163  0.017118  78.978602  15.744650   9.000000  1996.339515
202003  357.147860  0.019455  75.863813  13.311284   8.727626  1998.731518
202004  317.593023  0.046512  79.511628  17.005814   9.459302  1998.645349
202005  248.504132  0.041322  80.210744  14.020661   9.231405  1996.822314

In [ ]:
df.groupby("계약년월").median()

<ipython-input-129-d8f30dc145d7>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.median is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby("계약년월").median()


본번   부번  전용면적   계약일     층    건축년도
계약년월                                        
201906  122.0  0.0  84.0  15.0  10.0  1994.0
201907  133.0  0.0  84.0  17.0   9.0  1994.0
201908  176.0  0.0  84.0  17.0   9.0  1995.0
201909  143.0  0.0  84.0  19.0   8.0  1995.0
201910  132.0  0.0  84.0  16.0   9.0  1994.0
201911  133.0  0.0  84.0  16.0   8.0  1994.0
201912  180.0  0.0  84.0  11.0   8.5  1994.0
202001  201.0  0.0  68.0  14.0   9.0  1994.0
202002  192.0  0.0  70.0  16.0   8.0  1994.0
202003  212.0  0.0  59.0  12.0   8.0  1995.0
202004  181.0  0.0  59.0  18.0   9.0  1995.0
202005  133.0  0.0  66.5  13.5   8.0  1994.0

## DataFrame Data Handling

### Indexing, slicing

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   

             거래금액   층  건축년도      도로명  
0          49,000   2  1995  미금로 184  
1          73,200  15  1995  미금로 184  
2          51,000   6  1995  미금로 184  
3          51,000   2  1995  미금로 184  
4          54,000   6  1995  미금로 184

In [ ]:
df['단지명'][0]

'까치마을(1단지)(대우롯데선경)'

In [ ]:
df['시군구']

0       경기도 성남분당구 구미동
1       경기도 성남분당구 구미동
2       경기도 성남분당구 구미동
3       경기도 성남분당구 구미동
4       경기도 성남분당구 구미동
            ...      
7168    경기도 성남분당구 판교동
7169    경기도 성남분당구 판교동
7170    경기도 성남분당구 판교동
7171    경기도 성남분당구 판교동
7172    경기도 성남분당구 판교동
Name: 시군구, Length: 7173, dtype: object

In [ ]:
stock['2019-02-21': '2019-02-23']

2019-02-21    94300
2019-02-22    92100
2019-02-23    92400
dtype: int64

#### loc indexing, slicing

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   

             거래금액   층  건축년도      도로명  
0          49,000   2  1995  미금로 184  
1          73,200  15  1995  미금로 184  
2          51,000   6  1995  미금로 184  
3          51,000   2  1995  미금로 184  
4          54,000   6  1995  미금로 184

In [ ]:
df.loc[0, '시군구']

'경기도 성남분당구 구미동'

In [ ]:
df.loc[10, '거래금액']

'        85,000'

In [ ]:
df.loc[0:10, '거래금액']

0             49,000
1             73,200
2             51,000
3             51,000
4             54,000
5             51,250
6             54,600
7             52,000
8             87,000
9             89,500
10            85,000
Name: 거래금액, dtype: object

In [ ]:
df.loc[:, '전용면적']

0       51
1       70
2       51
3       51
4       51
        ..
7168    84
7169    71
7170    84
7171    71
7172    71
Name: 전용면적, Length: 7173, dtype: int64

#### iloc indexing, slicing

In [ ]:
df

시군구   번지   본번  부번                단지명  전용면적    계약년월  계약일  \
0     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    51  201906    1   
1     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
2     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    51  201906   10   
3     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
4     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
...             ...  ...  ...  ..                ...   ...     ...  ...   
7168  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   14   
7169  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202003   21   
7170  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   27   
7171  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202004   18   
7172  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202005    5   

                거래금액   층  건축년도      도로명  
0             49,000   2  1995  미금로 184  
1             73,200  15  1995  미금로 184  
2             51,000   6  1995  미금로 184  
3             51,000   2  1995  미금로 184  
4             54,000   6  1995  미금로 184  
...              ...  ..   ...      ...  
7168         101,000   1  2009  서판교로 29  
7169         100,500   6  2009  서판교로 29  
7170         100,200  16  2009  서판교로 29  
7171          97,000   9  2009  서판교로 29  
7172          98,000   3  2009  서판교로 29  

[7173 rows x 12 columns]

In [ ]:
df.iloc[100, 4]

'까치마을(2단지)(주공)'

In [ ]:
df.iloc[200, -1]

'미금일로 58'

In [ ]:
df.iloc[100:200, :]

시군구  번지  본번  부번            단지명  전용면적    계약년월  계약일  \
100  경기도 성남분당구 구미동  88  88   0  까치마을(2단지)(주공)    59  201912    7   
101  경기도 성남분당구 구미동  88  88   0  까치마을(2단지)(주공)    49  201912   11   
102  경기도 성남분당구 구미동  88  88   0  까치마을(2단지)(주공)    58  201912   16   
103  경기도 성남분당구 구미동  88  88   0  까치마을(2단지)(주공)    58  201912   17   
104  경기도 성남분당구 구미동  88  88   0  까치마을(2단지)(주공)    58  201912   18   
..             ...  ..  ..  ..            ...   ...     ...  ...   
195  경기도 성남분당구 구미동  63  63   0  까치마을(4단지)(롯데)    70  201907   16   
196  경기도 성남분당구 구미동  63  63   0  까치마을(4단지)(롯데)   101  201907   20   
197  경기도 성남분당구 구미동  63  63   0  까치마을(4단지)(롯데)    70  201907   23   
198  경기도 성남분당구 구미동  63  63   0  까치마을(4단지)(롯데)    70  201907   26   
199  경기도 성남분당구 구미동  63  63   0  까치마을(4단지)(롯데)    84  201907   26   

               거래금액   층  건축년도      도로명  
100          60,000   7  1995  미금일로 22  
101          52,500   2  1995  미금일로 22  
102          58,000  10  1995  미금일로 22  
103          56,000   1  1995  미금일로 22  
104          63,300   7  1995  미금일로 22  
..              ...  ..   ...      ...  
195          79,500  13  1995  미금일로 58  
196          98,800  10  1995  미금일로 58  
197          79,500  13  1995  미금일로 58  
198          70,000   2  1995  미금일로 58  
199          90,000  17  1995  미금일로 58  

[100 rows x 12 columns]

In [ ]:
df.iloc[:, 0:3]

시군구   번지   본번
0     경기도 성남분당구 구미동   77   77
1     경기도 성남분당구 구미동   77   77
2     경기도 성남분당구 구미동   77   77
3     경기도 성남분당구 구미동   77   77
4     경기도 성남분당구 구미동   77   77
...             ...  ...  ...
7168  경기도 성남분당구 판교동  585  585
7169  경기도 성남분당구 판교동  585  585
7170  경기도 성남분당구 판교동  585  585
7171  경기도 성남분당구 판교동  585  585
7172  경기도 성남분당구 판교동  585  585

[7173 rows x 3 columns]

#### 조건에 따라 추출하기

In [ ]:
df

시군구   번지   본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
0     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906    1   
1     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  70.6600  201906    7   
2     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   10   
3     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
4     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)  51.3200  201906   13   
...             ...  ...  ...  ..                ...      ...     ...  ...   
7168  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   14   
7169  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202003   21   
7170  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  84.9907  202003   27   
7171  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202004   18   
7172  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)  71.8438  202005    5   

      거래금액(만원)   층  건축년도      도로명  
0        49000   2  1995  미금로 184  
1        73200  15  1995  미금로 184  
2        51000   6  1995  미금로 184  
3        51000   2  1995  미금로 184  
4        54000   6  1995  미금로 184  
...        ...  ..   ...      ...  
7168    101000   1  2009  서판교로 29  
7169    100500   6  2009  서판교로 29  
7170    100200  16  2009  서판교로 29  
7171     97000   9  2009  서판교로 29  
7172     98000   3  2009  서판교로 29  

[7173 rows x 12 columns]

In [ ]:
df.loc[df['거래금액(만원)'] <90000  , "거래금액(만원)"]

count     4272.000000
mean     64886.455290
std      16349.853133
min      20600.000000
25%      50375.000000
50%      66500.000000
75%      79500.000000
max      89990.000000
Name: 거래금액(만원), dtype: float64

In [ ]:
df.loc[df['계약년월'] == 202003, :]

시군구   번지   본번  부번                단지명  전용면적    계약년월  계약일  \
56    경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    84  202003    5   
57    경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    51  202003    6   
123   경기도 성남분당구 구미동   88   88   0      까치마을(2단지)(주공)    59  202003    2   
187   경기도 성남분당구 구미동   66   66   0      까치마을(3단지)(신원)   101  202003   13   
238   경기도 성남분당구 구미동   63   63   0      까치마을(4단지)(롯데)    84  202003    6   
...             ...  ...  ...  ..                ...   ...     ...  ...   
7107  경기도 성남분당구 판교동  637  637   0  판교원마을7단지(모아미래도)임대    56  202003   14   
7108  경기도 성남분당구 판교동  637  637   0  판교원마을7단지(모아미래도)임대    80  202003   18   
7168  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   14   
7169  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202003   21   
7170  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   27   

                거래금액   층  건축년도       도로명  
56            98,000  13  1995   미금로 184  
57            65,000  10  1995   미금로 184  
123           67,700   8  1995   미금일로 22  
187          115,000  12  1995   미금로 177  
238          105,000  19  1995   미금일로 58  
...              ...  ..   ...       ...  
7107          88,500  11  2009  판교원로 237  
7108         103,000  11  2009  판교원로 237  
7168         101,000   1  2009   서판교로 29  
7169         100,500   6  2009   서판교로 29  
7170         100,200  16  2009   서판교로 29  

[257 rows x 12 columns]

In [ ]:
df.loc[(df['계약년월'] == 202003) & (df['전용면적'] > 59), :]

시군구   번지   본번  부번                 단지명  전용면적    계약년월  계약일  \
56    경기도 성남분당구 구미동   77   77   0   까치마을(1단지)(대우롯데선경)    84  202003    5   
187   경기도 성남분당구 구미동   66   66   0       까치마을(3단지)(신원)   101  202003   13   
238   경기도 성남분당구 구미동   63   63   0       까치마을(4단지)(롯데)    84  202003    6   
239   경기도 성남분당구 구미동   63   63   0       까치마을(4단지)(롯데)    84  202003   11   
283   경기도 성남분당구 구미동   63   63   0       까치마을(4단지)(선경)    84  202003   12   
...             ...  ...  ...  ..                 ...   ...     ...  ...   
7052  경기도 성남분당구 판교동  635  635   0  판교원마을6단지(진원로제비앙)임대    83  202003   19   
7108  경기도 성남분당구 판교동  637  637   0   판교원마을7단지(모아미래도)임대    80  202003   18   
7168  경기도 성남분당구 판교동  585  585   0     판교원마을9단지(한림풀에버)    84  202003   14   
7169  경기도 성남분당구 판교동  585  585   0     판교원마을9단지(한림풀에버)    71  202003   21   
7170  경기도 성남분당구 판교동  585  585   0     판교원마을9단지(한림풀에버)    84  202003   27   

                거래금액   층  건축년도       도로명  
56            98,000  13  1995   미금로 184  
187          115,000  12  1995   미금로 177  
238          105,000  19  1995   미금일로 58  
239          100,000  22  1995   미금일로 58  
283           95,000   1  1995   미금일로 58  
...              ...  ..   ...       ...  
7052          93,000   6  2009  판교원로 209  
7108         103,000  11  2009  판교원로 237  
7168         101,000   1  2009   서판교로 29  
7169         100,500   6  2009   서판교로 29  
7170         100,200  16  2009   서판교로 29  

[116 rows x 12 columns]

In [ ]:
df.loc[(df['계약년월'] == 202003) | (df['전용면적'] > 59), :]

시군구   번지   본번  부번                단지명  전용면적    계약년월  계약일  \
1     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
8     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    84  201906   28   
9     경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    84  201907    3   
10    경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    84  201907    4   
14    경기도 성남분당구 구미동   77   77   0  까치마을(1단지)(대우롯데선경)    70  201907   31   
...             ...  ...  ...  ..                ...   ...     ...  ...   
7168  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   14   
7169  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202003   21   
7170  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    84  202003   27   
7171  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202004   18   
7172  경기도 성남분당구 판교동  585  585   0    판교원마을9단지(한림풀에버)    71  202005    5   

                거래금액   층  건축년도      도로명  
1             73,200  15  1995  미금로 184  
8             87,000  10  1995  미금로 184  
9             89,500  19  1995  미금로 184  
10            85,000   2  1995  미금로 184  
14            78,000   6  1995  미금로 184  
...              ...  ..   ...      ...  
7168         101,000   1  2009  서판교로 29  
7169         100,500   6  2009  서판교로 29  
7170         100,200  16  2009  서판교로 29  
7171          97,000   9  2009  서판교로 29  
7172          98,000   3  2009  서판교로 29  

[4878 rows x 12 columns]

### DataFrame Data 생성/수정

##### 신규 데이터 칼럼 생성 및 데이터 입력(기본)

In [ ]:
df['mydata']=100

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70.66  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   

   거래금액(만원)   층  건축년도      도로명  mydata  
0     49000   2  1995  미금로 184     100  
1     73200  15  1995  미금로 184     100  
2     51000   6  1995  미금로 184     100  
3     51000   2  1995  미금로 184     100  
4     54000   6  1995  미금로 184     100

In [ ]:
df['계약년도'] = 2020

In [ ]:
df["동"] = df['시군구'].apply(lambda x: x.split(" ")[-1])

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적(㎡)    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70.66  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51.32  201906   13   

   거래금액(만원)   층  건축년도      도로명  mydata    동  
0     49000   2  1995  미금로 184     100  구미동  
1     73200  15  1995  미금로 184     100  구미동  
2     51000   6  1995  미금로 184     100  구미동  
3     51000   2  1995  미금로 184     100  구미동  
4     54000   6  1995  미금로 184     100  구미동

In [ ]:
df.loc[:, ['건축년도', '거래금액(만원)']].groupby("건축년도", as_index = False).mean()

건축년도       거래금액(만원)
0   1991   93603.325416
1   1992   90794.566879
2   1993   80255.909580
3   1994   75764.651618
4   1995   63962.178526
5   1996   67180.606061
6   1998   38523.076923
7   2003  115449.140893
8   2004  130970.058140
9   2005   71007.692308
10  2007  110897.435897
11  2008   82111.482759
12  2009  112088.929885
13  2010   97638.095238
14  2011  196078.947368
15  2012  180028.571429
16  2013  194583.333333
17  2015  170238.297872

In [ ]:
df.loc[:, ["계약년월", "동", "거래금액(만원)"]].groupby(by = ['계약년월', '동']).mean()

거래금액(만원)
계약년월   동                 
201906 구미동   58823.684211
       금곡동   66674.000000
       백현동  143625.000000
       분당동   72663.636364
       삼평동  126978.260870
...                   ...
202005 야탑동   57455.714286
       운중동   84050.181818
       이매동   90907.142857
       정자동   84944.680851
       판교동   94904.900000

[144 rows x 1 columns]

##### 함수를 이용하여 신규 데이터 입력하기 I

In [ ]:
df['계약년도'] = df['계약년월'].apply(lambda x: int(str(x)[:4]))

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   

             거래금액   층  건축년도      도로명  계약년도  
0          49,000   2  1995  미금로 184  2019  
1          73,200  15  1995  미금로 184  2019  
2          51,000   6  1995  미금로 184  2019  
3          51,000   2  1995  미금로 184  2019  
4          54,000   6  1995  미금로 184  2019

##### 함수를 이용하여 신규 데이터 입력하기 II

In [ ]:
df.head()

시군구  번지  본번  부번                단지명  전용면적    계약년월  계약일  \
0  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906    1   
1  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    70  201906    7   
2  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   10   
3  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   
4  경기도 성남분당구 구미동  77  77   0  까치마을(1단지)(대우롯데선경)    51  201906   13   

             거래금액   층  건축년도      도로명  계약년도  
0          49,000   2  1995  미금로 184  2019  
1          73,200  15  1995  미금로 184  2019  
2          51,000   6  1995  미금로 184  2019  
3          51,000   2  1995  미금로 184  2019  
4          54,000   6  1995  미금로 184  2019

In [ ]:
df['거래금액'].str.replace(",", "")

0               49000
1               73200
2               51000
3               51000
4               54000
            ...      
7168           101000
7169           100500
7170           100200
7171            97000
7172            98000
Name: 거래금액, Length: 7173, dtype: object

In [ ]:
df['거래금액'] = df['거래금액'].str.replace(",", "")

### 결측치(Missing Value Imputation) Handling

#### 결측치 확인

In [ ]:
import os

data_dir = 'data'
file_name = 'real_estate_price_edited.csv'
df_new = pd.read_csv( file_name)

In [ ]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    6939 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


#### 결측치 Drop

In [ ]:
df_copy = df_new.copy()

In [ ]:
## 결측치 확인
sum(df_copy['거래금액'].isnull())

548

In [ ]:
df_copy.isnull().sum(axis = 0)

시군구       0
번지        0
본번        0
부번        0
단지명       0
전용면적    234
계약년월      0
계약일       0
거래금액    548
층         0
건축년도      0
도로명       0
dtype: int64

##### 결측치가 존재하는 row를 기준으로 Drop

In [ ]:
df_copy.dropna(subset = ['거래금액'], inplace = True)

In [ ]:
## 결측치 확인
sum(df_copy['거래금액'].isnull())

0

In [ ]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6625 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     6625 non-null   object 
 1   번지      6625 non-null   object 
 2   본번      6625 non-null   int64  
 3   부번      6625 non-null   int64  
 4   단지명     6625 non-null   object 
 5   전용면적    6391 non-null   float64
 6   계약년월    6625 non-null   int64  
 7   계약일     6625 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       6625 non-null   int64  
 10  건축년도    6625 non-null   int64  
 11  도로명     6625 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.9+ KB


In [ ]:
df_copy.shape

(6625, 12)

##### 결측치가 존재하는 column을 기준으로 Drop

```python
DataFrame변수명.drop(columns = "Drop column명", inplace = True)
```

In [ ]:
df_copy.drop(columns = "전용면적", inplace = True)

In [ ]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6625 entries, 0 to 7172
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   시군구     6625 non-null   object
 1   번지      6625 non-null   object
 2   본번      6625 non-null   int64 
 3   부번      6625 non-null   int64 
 4   단지명     6625 non-null   object
 5   계약년월    6625 non-null   int64 
 6   계약일     6625 non-null   int64 
 7   거래금액    6625 non-null   object
 8   층       6625 non-null   int64 
 9   건축년도    6625 non-null   int64 
 10  도로명     6625 non-null   object
dtypes: int64(6), object(5)
memory usage: 621.1+ KB


#### 결측치 Fill

##### 일반적인 Data Fill 방식
- 직전 데이터, 직후 데이터로 채우기

In [ ]:
import os
data_path = './data'
file_name = 'kospi_data.csv'
kospi_df = pd.read_csv(os.path.join(data_path, file_name))

In [ ]:
kospi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744 entries, 0 to 743
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   date             744 non-null    object 
 1   close            668 non-null    float64
 2   volume           694 non-null    float64
 3   5day_MA_close    744 non-null    float64
 4   60day_MA_close   744 non-null    float64
 5   120day_MA_close  744 non-null    float64
 6   dollar_close     744 non-null    float64
 7   0                0 non-null      float64
dtypes: float64(7), object(1)
memory usage: 46.6+ KB


###### 이전값으로 채우기(before)

In [ ]:
kospi_df.head()

date    close    volume  5day_MA_close  60day_MA_close  \
0  2017-11-21  2530.70  484887.0      2529.1850     2529.185000   
1  2017-11-22  2540.51  485671.0      2532.9600     2532.960000   
2  2017-11-23  2537.15  391292.0      2534.0075     2534.007500   
3  2017-11-24  2544.33  297938.0      2536.0720     2536.072000   
4  2017-11-27  2507.81  284887.0      2532.1000     2531.361667   

   120day_MA_close  dollar_close   0  
0      2529.185000       1112.27 NaN  
1      2532.960000       1103.34 NaN  
2      2534.007500       1101.61 NaN  
3      2536.072000       1102.53 NaN  
4      2531.361667       1105.27 NaN

In [ ]:
kospi_df['close'] = kospi_df['close'].fillna(method = 'bfill')

<ipython-input-62-12bd71cda955>:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  kospi_df['close'] = kospi_df['close'].fillna(method = 'bfill')


In [ ]:
kospi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744 entries, 0 to 743
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   date             744 non-null    object 
 1   close            744 non-null    float64
 2   volume           694 non-null    float64
 3   5day_MA_close    744 non-null    float64
 4   60day_MA_close   744 non-null    float64
 5   120day_MA_close  744 non-null    float64
 6   dollar_close     744 non-null    float64
 7   0                0 non-null      float64
dtypes: float64(7), object(1)
memory usage: 46.6+ KB


###### 이후값으로 채우기(Forward)

In [ ]:
kospi_df['volume'] = kospi_df['volume'].fillna(method = 'ffill')

<ipython-input-64-e02278b8eb48>:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  kospi_df['volume'] = kospi_df['volume'].fillna(method = 'ffill')


In [ ]:
kospi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 744 entries, 0 to 743
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   date             744 non-null    object 
 1   close            744 non-null    float64
 2   volume           744 non-null    float64
 3   5day_MA_close    744 non-null    float64
 4   60day_MA_close   744 non-null    float64
 5   120day_MA_close  744 non-null    float64
 6   dollar_close     744 non-null    float64
 7   0                0 non-null      float64
dtypes: float64(7), object(1)
memory usage: 46.6+ KB


##### Numerical Data Fill
- 최빈값, 중앙값, 평균값 등의 방식이 있음.

###### 평균값 Fill

In [ ]:
df_copy_3 = df_new.copy()

In [ ]:
df_copy_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    6939 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


In [ ]:
fill_value = df_copy_3['전용면적'].median()
df_copy_3['전용면적'].fillna(fill_value, inplace = True)

<ipython-input-94-a13f75513887>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_copy_3['전용면적'].fillna(fill_value, inplace = True)


In [ ]:
df_copy_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    7173 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


###### 중앙값 Fill
- isnull함수와, notnull 함수를 활용하여 채우기

In [1]:
## null값이 포함된 칼럼은 바로 astype을 적용할 수 없다.
df_copy_3['거래금액'].str.replace(",", "").astype("int")

NameError: name 'df_copy_3' is not defined

In [ ]:
## 1. 거래금액 칼럼을 수정하기 위해, comma(,) 제거부터! but, nan 값을 float로 잡으니 str함수를 써서 일시적으로 str로 변환한 후에 replace, lstrip함수 적용
df_copy_3['거래금액'] = df_copy_3['거래금액'].str.replace(",", "").str.lstrip(" ")

In [ ]:
df_copy_3['거래금액'] = df_copy_3['거래금액'].astype("int")

In [ ]:
df_copy_3.loc[df_copy_3['거래금액'].isnull(), "거래금액"]

In [ ]:
df_copy_3.loc[df_copy_3['거래금액'].notnull(), "거래금액"].median()

In [ ]:
fill_value = df_copy_3.loc[df_copy_3['거래금액'].notnull(),"거래금액" ].astype("int").median()

In [ ]:
df_copy_3['거래금액'].fillna(fill_value, inplace = True)

In [ ]:
df_copy_3.info()

In [ ]:
## 3. 중앙값 Fill & Data Type 변경
df_copy_3['거래금액'] = df_copy_3['거래금액'].fillna(median_price).astype("int")

In [ ]:
df_copy_3.info()

##### Model Imputer
- Correlation 직접 계산 or KNN Impute 방식 사용

###### Correlation을 계산해서 인접 data 찾아 Missing Value Impute하기

In [ ]:
df_copy_4 = df_new.copy()

In [ ]:
df_copy_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    6939 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


In [ ]:
numeric_cols = df_copy_4.describe().columns
non_null_data = df_copy_4.loc[df_copy_4['전용면적'].notnull(), numeric_cols.drop('전용면적')]
null_data = df_copy_4.loc[df_copy_4['전용면적'].isnull(), numeric_cols.drop('전용면적')]

In [ ]:
x = null_data.iloc[0].values

In [ ]:
target_df = non_null_data

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
cosine_similarity(df_copy_4.loc[:, numeric_cols.drop('전용면적')])

array([[1.        , 1.        , 1.        , ..., 0.99999683, 0.99999683,
        0.99999684],
       [1.        , 1.        , 1.        , ..., 0.99999683, 0.99999683,
        0.99999683],
       [1.        , 1.        , 1.        , ..., 0.99999683, 0.99999684,
        0.99999684],
       ...,
       [0.99999683, 0.99999683, 0.99999683, ..., 1.        , 1.        ,
        0.99999999],
       [0.99999683, 0.99999683, 0.99999684, ..., 1.        , 1.        ,
        1.        ],
       [0.99999684, 0.99999683, 0.99999684, ..., 0.99999999, 1.        ,
        1.        ]])

In [ ]:
cos_sim_arr = cosine_similarity(df_copy_4.loc[:, numeric_cols.drop('전용면적')])

In [ ]:
cos_sim_arr[null_data.index]

array([[0.99999983, 0.99999984, 0.99999984, ..., 0.99999811, 0.99999811,
        0.99999811],
       [0.99999983, 0.99999983, 0.99999984, ..., 0.99999811, 0.99999811,
        0.9999981 ],
       [0.99999983, 0.99999983, 0.99999983, ..., 0.99999811, 0.99999811,
        0.9999981 ],
       ...,
       [0.99999996, 0.99999996, 0.99999996, ..., 0.99999749, 0.99999749,
        0.99999749],
       [0.99999996, 0.99999996, 0.99999996, ..., 0.99999749, 0.99999749,
        0.99999749],
       [0.99999996, 0.99999996, 0.99999996, ..., 0.99999749, 0.99999749,
        0.9999975 ]])

In [ ]:
for idx in df_copy_4.loc[df_copy_4['전용면적'].isnull()].index:
  arg_sort_list =  np.argsort(cos_sim_arr[idx]).tolist()
  if arg_sort_list[-1] == idx:
    nn_index_list = arg_sort_list[-4:-1]
  else:
    nn_index_list = arg_sort_list[-3:]


  df_copy_4.loc[idx, '전용면적'] = df_copy_4.loc[nn_index_list, '전용면적'].mean()

In [ ]:
df_copy_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   int64  
 3   부번      7173 non-null   int64  
 4   단지명     7173 non-null   object 
 5   전용면적    7114 non-null   float64
 6   계약년월    7173 non-null   int64  
 7   계약일     7173 non-null   int64  
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   int64  
 10  건축년도    7173 non-null   int64  
 11  도로명     7173 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 672.6+ KB


###### KNN Imputer (지금은 참고만..)

In [ ]:
from sklearn.impute import KNNImputer

In [ ]:
df_copy_5 = df_new.copy()
numeric_cols = df_copy_5.describe().columns
imputer = KNNImputer(n_neighbors = 5)
imputer.fit(df_copy_5.loc[:, numeric_cols])

KNNImputer()

In [ ]:
df_copy_5.loc[:, numeric_cols] = imputer.transform(df_copy_5.loc[:, numeric_cols])

<ipython-input-139-09e54898e6cb>:1: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  df_copy_5.loc[:, numeric_cols] = imputer.transform(df_copy_5.loc[:, numeric_cols])


In [ ]:
df_copy_5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173 entries, 0 to 7172
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   시군구     7173 non-null   object 
 1   번지      7173 non-null   object 
 2   본번      7173 non-null   float64
 3   부번      7173 non-null   float64
 4   단지명     7173 non-null   object 
 5   전용면적    7173 non-null   float64
 6   계약년월    7173 non-null   float64
 7   계약일     7173 non-null   float64
 8   거래금액    6625 non-null   object 
 9   층       7173 non-null   float64
 10  건축년도    7173 non-null   float64
 11  도로명     7173 non-null   object 
dtypes: float64(7), object(5)
memory usage: 672.6+ KB


#### Categorical Data Fill

##### 최빈값 Fill
- 최빈값은 mode함수를 사용해 구함.

In [ ]:
import pandas as pd

In [ ]:
salary_df = pd.read_csv("./ds_salaries.csv")

In [ ]:
salary_df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [ ]:
salary_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3730 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [ ]:
salary_df['job_title'].value_counts()

Data Engineer                          1030
Data Scientist                          835
Data Analyst                            608
Machine Learning Engineer               289
Analytics Engineer                      103
                                       ... 
Principal Machine Learning Engineer       1
Azure Data Engineer                       1
Manager Data Management                   1
Marketing Data Engineer                   1
Finance Data Analyst                      1
Name: job_title, Length: 93, dtype: int64

In [ ]:
salary_df['job_title'].mode()

0    Data Engineer
Name: job_title, dtype: object

In [ ]:
salary_df['job_title'] =  salary_df['job_title'].fillna( salary_df['job_title'].mode()[0])

In [ ]:
salary_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


### Dummy Variable 생성(One-Hot Encoding)

#### object Data확인

In [ ]:
salary_df.select_dtypes("object").columns

Index(['experience_level', 'employment_type', 'job_title', 'salary_currency',
       'employee_residence', 'company_location', 'company_size'],
      dtype='object')

In [ ]:
salary_df['experience_level'].unique()

array(['SE', 'MI', 'EN', 'EX'], dtype=object)

In [ ]:
pd.get_dummies(salary_df, columns = ['experience_level'])

work_year employment_type                 job_title   salary  \
0          2023              FT  Principal Data Scientist    80000   
1          2023              CT               ML Engineer    30000   
2          2023              CT               ML Engineer    25500   
3          2023              FT            Data Scientist   175000   
4          2023              FT            Data Scientist   120000   
...         ...             ...                       ...      ...   
3750       2020              FT            Data Scientist   412000   
3751       2021              FT  Principal Data Scientist   151000   
3752       2020              FT            Data Scientist   105000   
3753       2020              CT     Business Data Analyst   100000   
3754       2021              FT      Data Science Manager  7000000   

     salary_currency  salary_in_usd employee_residence  remote_ratio  \
0                EUR          85847                 ES           100   
1                USD          30000                 US           100   
2                USD          25500                 US           100   
3                USD         175000                 CA           100   
4                USD         120000                 CA           100   
...              ...            ...                ...           ...   
3750             USD         412000                 US           100   
3751             USD         151000                 US           100   
3752             USD         105000                 US           100   
3753             USD         100000                 US           100   
3754             INR          94665                 IN            50   

     company_location company_size  experience_level_EN  experience_level_EX  \
0                  ES            L                False                False   
1                  US            S                False                False   
2                  US            S                False                False   
3                  CA            M                False                False   
4                  CA            M                False                False   
...               ...          ...                  ...                  ...   
3750               US            L                False                False   
3751               US            L                False                False   
3752               US            S                 True                False   
3753               US            L                 True                False   
3754               IN            L                False                False   

      experience_level_MI  experience_level_SE  
0                   False                 True  
1                    True                False  
2                    True                False  
3                   False                 True  
4                   False                 True  
...                   ...                  ...  
3750                False                 True  
3751                 True                False  
3752                False                False  
3753                False                False  
3754                False                 True  

[3755 rows x 14 columns]

In [ ]:
salary_df = pd.get_dummies(salary_df, columns = ['experience_level'])

#### object data 전부 One hot encoding 처리

In [ ]:
salary_df = pd.get_dummies(salary_df, columns = salary_df.select_dtypes("object").columns)

In [ ]:
salary_df.columns

Index(['work_year', 'salary', 'salary_in_usd', 'remote_ratio',
       'experience_level_EN', 'experience_level_EX', 'experience_level_MI',
       'experience_level_SE', 'employment_type_CT', 'employment_type_FL',
       ...
       'company_location_SI', 'company_location_SK', 'company_location_TH',
       'company_location_TR', 'company_location_UA', 'company_location_US',
       'company_location_VN', 'company_size_L', 'company_size_M',
       'company_size_S'],
      dtype='object', length=278)

In [ ]:
salary_df.shape

(3755, 278)

column개수가 278개인 거대 vector가 생성되었습니다.  
remote_ratio이후의 칼럼들은 대부분 0으로 채워지고, 일부만 1로 채워진 <strong> "Sparse" </strong> 한 Vector입니다.  
이런 Sparse한 Vector는 이른바 "차원의 저주"를 만들어내며, ML 학습을 방해하는 요소로 작용할 수 있습니다.  
  
따라서, 과도한 차원의 증가를 막기 위해, ML에서는 Embedding, PCA 등의 차원축소 기법을 사용하기도 합니다.  
혹은, 유사 그룹들을 묶어주는 Feature engineering작업을 수행하기도 합니다.

### Data 합치기

#### concat
- axis 0 : row-wise로 concat
- axis 1 : column-wise로 concat


In [ ]:
salary_df = pd.read_csv('./ds_salaries.csv')

In [ ]:
salary_df.shape

(3755, 11)

In [ ]:
salary_df['company_size'].unique().shape

(3,)

In [ ]:
salary_df_company_size = pd.get_dummies(salary_df['company_size'], prefix = 'company_size')

In [ ]:
salary_df_company_size.shape

(3755, 3)

In [ ]:
salary_df_new = pd.concat([salary_df, salary_df_company_size], axis = 1)

In [ ]:
salary_df_new.shape

(7510, 14)

In [ ]:
salary_df_new.head()

work_year experience_level employment_type                 job_title  \
0     2023.0               SE              FT  Principal Data Scientist   
1     2023.0               MI              CT               ML Engineer   
2     2023.0               MI              CT               ML Engineer   
3     2023.0               SE              FT            Data Scientist   
4     2023.0               SE              FT            Data Scientist   

     salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000.0             EUR        85847.0                 ES         100.0   
1   30000.0             USD        30000.0                 US         100.0   
2   25500.0             USD        25500.0                 US         100.0   
3  175000.0             USD       175000.0                 CA         100.0   
4  120000.0             USD       120000.0                 CA         100.0   

  company_location company_size  company_size_L  company_size_M  \
0               ES            L             NaN             NaN   
1               US            S             NaN             NaN   
2               US            S             NaN             NaN   
3               CA            M             NaN             NaN   
4               CA            M             NaN             NaN   

   company_size_S  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

#### merge

In [ ]:
sales_df = pd.read_csv('./sales_info.csv')

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xbb in position 3: invalid start byte

In [ ]:
# csv파일을 읽어올때, decode error, encoding error 등이 발생하면, encoding type을 바꿔줍니다.
# encoding type은 최초에 파일을 저장했던 OS의 encoding type을 따라갑니다.
# 최초에 저장했던 OS의 encoding타입과 현재 내가 사용하고있는 OS의 encoding type이 맞지 않을경우 에러가 발생합니다.
# windows계열은 encoding type을 'cp949'를 사용하고, linux/unix계열은 utf-8을 사용합니다.

sales_df = pd.read_csv('./sales_info.csv', encoding = 'cp949')

In [ ]:
user_df = pd.read_csv('./user_info.csv', encoding = 'utf-8')

In [ ]:
sales_df.head()

ID 상품대분류명     상품중분류명    구매지역      구매일자  구매시간  구매수량    구매금액  취소여부
0  410362886   의류잡화  여성용의류-정장류  서울 송파구  20140510     1     1  198000     0
1  643279402     식품        농산물  경기 성남시  20140612     1     1    2000     0
2  643279402     식품        농산물  경기 성남시  20140612     1     1    4000     0
3  643279402     식품        농산물  경기 성남시  20140612     1     1    5000     0
4  643279402     식품        농산물  경기 성남시  20140612     1     1    9000     0

In [ ]:
user_df.head()

ID   성별    연령     거주지역
0  478207946.0  1.0  84.0   서울 성동구
1  479806984.0  1.0  84.0   서울 서초구
2   94790213.0  1.0  84.0   부산 사상구
3  656026338.0  1.0  84.0  서울 영등포구
4  433076833.0  1.0  83.0   경기 고양시

In [ ]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3590 entries, 0 to 3589
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      3590 non-null   float64
 1   성별      3590 non-null   float64
 2   연령      3590 non-null   float64
 3   거주지역    3590 non-null   object 
dtypes: float64(3), object(1)
memory usage: 112.3+ KB


In [ ]:
from random import randint

In [ ]:
sales_df.shape

(101692, 9)

In [ ]:
user_df.shape

(3590, 4)

In [ ]:
sales_df['ID'].unique().shape, user_df['ID'].unique().shape

((3542,), (3590,))

##### inner join
```python
DataFrame변수명 = pd.merge(DataFrame_1, DataFrame_2, on = 기준key값, how = 'inner')
```

In [ ]:
new_df = pd.merge(sales_df, user_df, on = 'ID', how = 'inner')

In [ ]:
new_df.head()

ID 상품대분류명     상품중분류명    구매지역      구매일자  구매시간  구매수량    구매금액  취소여부  \
0  410362886   의류잡화  여성용의류-정장류  서울 송파구  20140510     1     1  198000     0   
1  643279402     식품        농산물  경기 성남시  20140612     1     1    2000     0   
2  643279402     식품        농산물  경기 성남시  20140612     1     1    4000     0   
3  643279402     식품        농산물  경기 성남시  20140612     1     1    5000     0   
4  643279402     식품        농산물  경기 성남시  20140612     1     1    9000     0   

    성별    연령    거주지역  
0  1.0  69.0  부산 영도구  
1  2.0  56.0  경기 성남시  
2  2.0  56.0  경기 성남시  
3  2.0  56.0  경기 성남시  
4  2.0  56.0  경기 성남시

In [ ]:
new_df.shape
## DataFrame의 shape을 찍어보면 size가 줄어든 것을 확인할 수 있다.

(101686, 12)

In [ ]:
new_df.shape

(101686, 12)

In [ ]:
sales_df.shape

(101692, 9)

In [ ]:
print("Dataframe_1 과 Merge된 DataFrame 간 Key값 비교 : ", set(sales_df['ID']) -  set(new_df['ID']))
print("Dataframe_1 과 DataFrame_2 간 Key값 비교       : ", set(sales_df['ID']) -  set(user_df['ID'].astype("int")))

Dataframe_1 과 Merge된 DataFrame 간 Key값 비교 :  {148534985}
Dataframe_1 과 DataFrame_2 간 Key값 비교       :  {148534985}


##### left outer join
```python
DataFrame변수명 = pd.merge(DataFrame_1, DataFrame_2, on = 기준key값, how = 'left')
```

In [ ]:
new_df_2 = pd.merge(sales_df, user_df, left_on = 'ID',right_on = 'ID', how = 'left')

In [ ]:
new_df_2.head()

ID 상품대분류명     상품중분류명    구매지역      구매일자  구매시간  구매수량    구매금액  취소여부  \
0  410362886   의류잡화  여성용의류-정장류  서울 송파구  20140510     1     1  198000     0   
1  643279402     식품        농산물  경기 성남시  20140612     1     1    2000     0   
2  643279402     식품        농산물  경기 성남시  20140612     1     1    4000     0   
3  643279402     식품        농산물  경기 성남시  20140612     1     1    5000     0   
4  643279402     식품        농산물  경기 성남시  20140612     1     1    9000     0   

    성별    연령    거주지역  
0  1.0  69.0  부산 영도구  
1  2.0  56.0  경기 성남시  
2  2.0  56.0  경기 성남시  
3  2.0  56.0  경기 성남시  
4  2.0  56.0  경기 성남시

In [ ]:
new_df_2.shape
## DataFrame의 shape을 찍어보면 row size가 sales_df와 동일한 것을 확인할 수 있다.

(101692, 12)

In [ ]:
print("Dataframe_1 과 Merge된 DataFrame 간 Key값 비교 : ", set(sales_df['ID']) -  set(new_df_2['ID']))
print("Dataframe_1 과 DataFrame_2 간 Key값 비교       : ", set(sales_df['ID']) -  set(user_df['ID'].astype("int")))

Dataframe_1 과 Merge된 DataFrame 간 Key값 비교 :  set()
Dataframe_1 과 DataFrame_2 간 Key값 비교       :  {148534985}


##### Full outer join
```python
DataFrame변수명 = pd.merge(DataFrame_1, DataFrame_2, on = 기준key값, how = 'outer')
```

In [ ]:
new_df_3 = pd.merge(sales_df, user_df, on = 'ID', how = 'outer')

In [ ]:
new_df_3.head()

ID 상품대분류명     상품중분류명     구매지역        구매일자  구매시간  구매수량      구매금액  \
0  410362886.0   의류잡화  여성용의류-정장류   서울 송파구  20140510.0   1.0   1.0  198000.0   
1  410362886.0     식품       가공식품  서울 영등포구  20140809.0  10.0   1.0    3000.0   
2  410362886.0     식품       가공식품  서울 영등포구  20141004.0  10.0   2.0   12000.0   
3  410362886.0     식품        농산물  서울 영등포구  20140809.0  10.0   1.0    1000.0   
4  410362886.0     식품        수산물  서울 영등포구  20141004.0  10.0   1.0   13000.0   

   취소여부   성별    연령    거주지역  
0   0.0  1.0  69.0  부산 영도구  
1   0.0  1.0  69.0  부산 영도구  
2   0.0  1.0  69.0  부산 영도구  
3   0.0  1.0  69.0  부산 영도구  
4   0.0  1.0  69.0  부산 영도구

In [ ]:
sales_df.shape

(101692, 9)

In [ ]:
new_df_3.shape
## DataFrame의 shape을 찍어보면 row size가 sales_df보다 늘어난 것을 확인할 수 있다.

(101741, 12)

In [ ]:
new_df_3.shape[0] - sales_df.shape[0]

49

In [ ]:
print("Merge된 DataFrame과 Dataframe_1간 Key값 비교 : ", set(new_df_3['ID']) - set(sales_df['ID']) )
print("Merge된 DataFrame과 Dataframe_1간 Key값 비교 : ", len(set(new_df_3['ID']) - set(sales_df['ID']) ))
print("Dataframe_1 과 DataFrame_2 간 Key값 비교       : ", set(sales_df['ID']) -  set(user_df['ID'].astype("int")))

Merge된 DataFrame과 Dataframe_1간 Key값 비교 :  {1379714.0, 1244164.0, 1980037.0, 1465350.0, 1725200.0, 1076375.0, 1841687.0, 1609369.0, 1270047.0, 1184799.0, 1399968.0, 1842848.0, 1878307.0, 1590180.0, 1794087.0, 1583915.0, 1473707.0, 1244587.0, 1484974.0, 1289141.0, 1635125.0, 1937855.0, 1654085.0, 148534986.0, 1540428.0, 1614159.0, 1442128.0, 1516753.0, 1128913.0, 1682131.0, 1697877.0, 1883869.0, 1103200.0, 1047648.0, 1432546.0, 1959139.0, 1352167.0, 1181031.0, 1328361.0, 1192553.0, 1495657.0, 1881961.0, 1177707.0, 1293549.0, 1767538.0, 1708403.0, 1911156.0, 1376115.0, 1151994.0}
Merge된 DataFrame과 Dataframe_1간 Key값 비교 :  49
Dataframe_1 과 DataFrame_2 간 Key값 비교       :  {148534985}


### Variable Screening


#### Feature Selection
- mRMR은 classical statistic에 기반한 방식이고, Boruta나 All but X와 같은 방식들은 Model에 기반한 방식입니다.

##### mRMR(minimum Redundancy Maximum Relevancy)

In [ ]:
import pandas as pd

In [ ]:
salary_df = pd.read_csv('./data/ds_salaries.csv')

In [ ]:
salary_df.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

In [ ]:
salary_df = pd.get_dummies(salary_df, columns = ['company_size', 'experience_level', 'employment_type'])

In [ ]:
numeric_cols = salary_df.describe().columns

In [ ]:
numeric_df = salary_df.loc[:, numeric_cols]

In [ ]:
target_col = 'salary'
target_data = salary_df.loc[:, [target_col]]

In [ ]:
!pip install mrmr_selection

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 2.6 MB/s eta 0:00:00


In [ ]:
import mrmr
x = numeric_df
y = target_data

In [ ]:
selected_features = mrmr.mrmr_regression(X = x, y = y, K = 7)

100%|██████████| 7/7 [00:00<00:00, 16.49it/s]


In [ ]:
selected_features

['company_size_L',
 'experience_level_EX',
 'work_year',
 'company_size_M',
 'experience_level_MI',
 'remote_ratio',
 'experience_level_SE']

In [ ]:
numeric_df.columns

Index(['work_year', 'salary', 'salary_in_usd', 'remote_ratio',
       'company_size_L', 'company_size_M', 'company_size_S',
       'experience_level_EN', 'experience_level_EX', 'experience_level_MI',
       'experience_level_SE', 'employment_type_CT', 'employment_type_FL',
       'employment_type_FT', 'employment_type_PT'],
      dtype='object')

##### Boruta : 다중공선성은 고려하지 않고 과적합만을 고려하기 위한 방법입니다.

In [ ]:
target_col = 'salary'
x = numeric_df.loc[:, numeric_cols.drop(target_col)]
y = numeric_df.loc[:, target_col]

In [ ]:
!pip install boruta

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.6/56.6 kB 1.6 MB/s eta 0:00:00


In [ ]:
from boruta import BorutaPy
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor()
boruta = BorutaPy(estimator = model, n_estimators = 'auto', max_iter = 200)
boruta.fit(x.values, y.values)

BorutaPy(estimator=RandomForestRegressor(n_estimators=14,
                                         random_state=RandomState(MT19937) at 0x7F04407B7E40),
         max_iter=200, n_estimators='auto',
         random_state=RandomState(MT19937) at 0x7F04407B7E40)

In [ ]:
x.columns[boruta.support_]

Index(['salary_in_usd'], dtype='object')